File: WINGS_AI_Bootcamp/Elixir/Anagram_list_sowpods.exs
defmodule Anagram do
  def anagram_list(file_path) do
    stream = File.stream!(file_path)
    trimmed = Stream.map(stream, fn line -> String.trim(line) end)
    lowered = Stream.map(trimmed, fn word -> String.downcase(word) end)
    words = Enum.to_list(lowered)

    anagram_map = Enum.reduce(words, %{}, fn word, acc ->
      graphemes = String.graphemes(word)
      sorted_graphemes = Enum.sort(graphemes)
      key = Enum.join(sorted_graphemes)
      Map.update(acc, key, [word], fn existing -> [word | existing] end)
    end)

    Enum.each(anagram_map, fn {_key, group} ->
      if length(group) > 1 do
        sorted_group = Enum.sort(group)
        line = Enum.join(sorted_group, " ")
        IO.puts(line)
      end
    end)
  end
end

# Call the function
Anagram.anagram_list("sowpods.txt")

File: WINGS_AI_Bootcamp/Haskell/Anagram_list_sowpods.hs
import Data.List (sort, groupBy, sortOn)
import Data.Char (toLower)
import System.IO

-- Generate the key by sorting lowercase letters of the word
key :: String -> String
key = sort . map toLower

anagramList :: FilePath -> IO ()
anagramList filePath = do
    contents <- readFile filePath
    let wordsList = lines contents
        -- Pair each word with its key
        paired = [(key w, w) | w <- wordsList]
        -- Sort by key to group anagrams together
        sortedPairs = sortOn fst paired
        -- Group by key
        grouped = groupBy (\a b -> fst a == fst b) sortedPairs
        -- Filter groups with more than one word
        filtered = filter (\g -> length g > 1) grouped
        -- Extract only words and sort each group
        anagramGroups = map (sort . map snd) filtered
    -- Print each group on its own line
    mapM_ (putStrLn . unwords) anagramGroups

main :: IO ()
main = anagramList "sowpods.txt"

